File: frontend/src/services/api/dataService.ts
/**
 * Data Service - Handles all data-related API calls to the backend
 * Clean abstraction layer for backend communication
 */

export interface QueryRequest {
  query: string;
}

export interface QueryResult {
  data: any[];
  rows: number;
  columns: string[];
}

export interface CorrelationRequest {
  symbols: string[];
  exchange?: string;
}

export interface CorrelationResult {
  correlation: number | null;
  symbol1_stats: Record<string, any>;
  symbol2_stats: Record<string, any>;
}

export interface DataSummary {
  total_bars: number;
  symbols: Array<{
    symbol: string;
    exchange: string;
    bar_count: number;
    first_bar: string;
    last_bar: string;
  }>;
}

export interface SaveDataRequest {
  symbol: string;
  exchange: string;
  candles: number[][];
  interval: string;
}

class DataService {
  private baseUrl = 'http://localhost:5001';

  /**
   * Get summary of all stored Parquet data
   */
  async getDataSummary(): Promise<DataSummary> {
    const response = await fetch(`${this.baseUrl}/api/data/summary`);
    
    if (!response.ok) {
      throw new Error(`Failed to fetch data summary: ${response.statusText}`);
    }
    
    return response.json();
  }

  /**
   * Execute SQL query on the backend DuckDB
   */
  async queryData(query: string): Promise<QueryResult> {
    const response = await fetch(`${this.baseUrl}/api/data/query`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ query }),
    });

    if (!response.ok) {
      const error = await response.json().catch(() => ({ error: response.statusText }));
      throw new Error(error.error || `Query failed: ${response.statusText}`);
    }

    return response.json();
  }

  /**
   * Get correlation between two symbols
   */
  async getCorrelation(symbol1: string, symbol2: string): Promise<CorrelationResult> {
    // Convert internal format (BTC/USD) to URL format (BTC-USD)
    const urlSymbol1 = symbol1.replace('/', '-');
    const urlSymbol2 = symbol2.replace('/', '-');
    
    const response = await fetch(`${this.baseUrl}/api/data/correlation/${urlSymbol1}/${urlSymbol2}`);
    
    if (!response.ok) {
      throw new Error(`Failed to get correlation: ${response.statusText}`);
    }
    
    return response.json();
  }

  /**
   * Get correlation matrix for multiple symbols
   */
  async getCorrelationMatrix(symbols: string[], exchange = 'coinbase'): Promise<any> {
    const response = await fetch(`${this.baseUrl}/api/analysis/correlation-matrix`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ symbols, exchange }),
    });

    if (!response.ok) {
      const error = await response.json().catch(() => ({ error: response.statusText }));
      throw new Error(error.error || `Correlation matrix failed: ${response.statusText}`);
    }

    return response.json();
  }

  /**
   * Save market data (for caching)
   */
  async saveMarketData(data: SaveDataRequest): Promise<{ status: string; message: string }> {
    const response = await fetch(`${this.baseUrl}/api/market-data/save`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(data),
    });

    if (!response.ok) {
      throw new Error(`Failed to save market data: ${response.statusText}`);
    }

    return response.json();
  }

  /**
   * Proxy request to Coinbase API (with automatic Parquet saving)
   */
  async proxyCoinbase(endpoint: string, params: Record<string, string> = {}): Promise<any> {
    const url = new URL(`${this.baseUrl}/api/proxy/coinbase/${endpoint}`);
    Object.entries(params).forEach(([key, value]) => {
      url.searchParams.append(key, value);
    });

    const response = await fetch(url.toString());

    if (!response.ok) {
      throw new Error(`Coinbase proxy failed: ${response.statusText}`);
    }

    return response.json();
  }

  /**
   * Get catalog of available data files
   */
  async getCatalogData(): Promise<any> {
    const response = await fetch(`${this.baseUrl}/api/catalog/list`);
    
    if (!response.ok) {
      throw new Error(`Failed to fetch catalog: ${response.statusText}`);
    }
    
    return response.json();
  }
}

// Export singleton instance
export const dataService = new DataService();

File: backend/core/data_service.py
"""
Data Service - Handles all data operations
Business logic layer for market data management
"""
from typing import Dict, Any, Optional, List
from pathlib import Path
import json
import requests
from data_manager import DataManager


class DataService:
    """Service layer for data operations"""
    
    def __init__(self, data_manager: DataManager = None):
        self.data_manager = data_manager or DataManager()
    
    def save_market_data(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Save market data from frontend IndexedDB cache"""
        try:
            symbol = data.get('symbol')
            exchange = data.get('exchange')
            candles = data.get('candles', [])
            
            # Log the save request
            print(f"📊 Market data save request: {symbol} from {exchange}, {len(candles)} candles")
            
            # Actually save the data to Parquet/DuckDB
            if candles:
                save_result = self.data_manager.save_coinbase_data(candles, symbol, exchange)
                print(f"✅ Saved to Parquet: {save_result}")
                
                return {
                    'status': 'success',
                    'message': f'Saved {save_result.get("bars_saved", 0)} candles for {symbol}',
                    'symbol': symbol,
                    'exchange': exchange,
                    'candle_count': save_result.get('bars_saved', 0),
                    'parquet_path': save_result.get('parquet_path', ''),
                    'date_range': save_result.get('date_range', {})
                }
            else:
                return {
                    'status': 'error',
                    'message': 'No candles data provided'
                }
            
        except Exception as e:
            print(f"❌ Error saving market data: {str(e)}")
            return {
                'status': 'error',
                'message': str(e)
            }
    
    def proxy_coinbase_data(self, endpoint: str, params: Dict[str, Any]) -> Dict[str, Any]:
        """Proxy requests to Coinbase API and save to Parquet"""
        try:
            # Build the Coinbase API URL
            base_url = 'https://api.exchange.coinbase.com'
            url = f"{base_url}/{endpoint}"
            
            # Make the request to Coinbase
            cb_response = requests.get(url, params=params, timeout=10)
            
            # If this is a candles request, save to Parquet
            if 'candles' in endpoint and cb_response.status_code == 200:
                try:
                    # Extract symbol from endpoint (e.g., "products/BTC-USD/candles" -> "BTC-USD")
                    parts = endpoint.split('/')
                    if len(parts) >= 2 and parts[0] == 'products':
                        symbol = parts[1].replace('-', '/')  # Convert BTC-USD to BTC/USD
                        
                        # Save to Parquet via DataManager
                        candles_data = cb_response.json()
                        if candles_data:
                            save_result = self.data_manager.save_coinbase_data(candles_data, symbol)
                            print(f"📊 Saved {save_result.get('bars_saved', 0)} bars for {symbol} to Parquet")
                except Exception as e:
                    print(f"⚠️ Failed to save candles to Parquet: {e}")
            
            return {
                'status': 'success',
                'data': cb_response.json(),
                'status_code': cb_response.status_code
            }
            
        except requests.exceptions.Timeout:
            return {
                'status': 'error',
                'message': 'Request to Coinbase timed out',
                'status_code': 504
            }
        except requests.exceptions.RequestException as e:
            return {
                'status': 'error',
                'message': f'Error proxying to Coinbase: {str(e)}',
                'status_code': 502
            }
        except Exception as e:
            return {
                'status': 'error',
                'message': f'Unexpected error: {str(e)}',
                'status_code': 500
            }
    
    def proxy_kraken_data(self, endpoint: str, params: Dict[str, Any]) -> Dict[str, Any]:
        """Proxy requests to Kraken API to avoid CORS issues"""
        try:
            import requests
            
            # Construct Kraken API URL
            base_url = "https://api.kraken.com"
            url = f"{base_url}/{endpoint}"
            
            # Make request to Kraken
            response = requests.get(url, params=params, timeout=10)
            
            if response.status_code == 200:
                data = response.json()
                return {
                    'status': 'success',
                    'data': data,
                    'status_code': 200
                }
            else:
                return {
                    'status': 'error',
                    'message': f'Kraken API error: {response.status_code}',
                    'status_code': response.status_code
                }
        except requests.exceptions.Timeout:
            return {
                'status': 'error',
                'message': 'Kraken API request timeout',
                'status_code': 504
            }
        except Exception as e:
            return {
                'status': 'error',
                'message': f'Unexpected error: {str(e)}',
                'status_code': 500
            }
    
    def get_data_summary(self) -> Dict[str, Any]:
        """Get summary of all stored Parquet data"""
        try:
            return self.data_manager.get_summary()
        except Exception as e:
            raise Exception(f"Failed to get data summary: {str(e)}")
    
    def query_data(self, query: str) -> Dict[str, Any]:
        """Execute SQL query on DuckDB"""
        try:
            # Safety check - only allow SELECT queries
            if not query.strip().upper().startswith('SELECT'):
                raise ValueError('Only SELECT queries are allowed')
            
            df = self.data_manager.query(query)
            result = df.to_dict(orient='records')
            
            return {
                'data': result,
                'rows': len(result),
                'columns': list(df.columns)
            }
        except Exception as e:
            raise Exception(f"Query execution failed: {str(e)}")
    
    def get_correlation(self, symbol1: str, symbol2: str) -> Dict[str, Any]:
        """Get correlation between two symbols"""
        try:
            # Convert URL format (BTC-USD) to our format (BTC/USD)
            symbol1 = symbol1.replace('-', '/')
            symbol2 = symbol2.replace('-', '/')
            
            correlation = self.data_manager.calculate_correlation(symbol1, symbol2)
            stats1 = self.data_manager.calculate_statistics(symbol1)
            stats2 = self.data_manager.calculate_statistics(symbol2)
            
            return {
                'correlation': correlation,
                'symbol1_stats': stats1,
                'symbol2_stats': stats2
            }
        except Exception as e:
            raise Exception(f"Correlation calculation failed: {str(e)}")
    
    def list_catalog_data(self) -> Dict[str, Any]:
        """List all available data in the backend catalog"""
        try:
            catalog_path = Path(__file__).parent.parent / 'catalog' / 'data'
            result = {
                'bars': [],
                'quotes': [],
                'trades': [],
                'signals': [],
                'backtests': []
            }
            
            if catalog_path.exists():
                # List bar data files
                bar_path = catalog_path / 'bar'
                if bar_path.exists():
                    result['bars'] = [
                        {
                            'filename': f.name,
                            'symbol': f.stem.split('-')[0] if '-' in f.stem else f.stem,
                            'timeframe': f.stem.split('-')[1] if '-' in f.stem else 'unknown',
                            'size': f.stat().st_size,
                            'modified': f.stat().st_mtime
                        }
                        for f in bar_path.glob('*.parquet')
                    ]
                
                # List quote data files
                quote_path = catalog_path / 'quote'
                if quote_path.exists():
                    result['quotes'] = [
                        {
                            'filename': f.name,
                            'symbol': f.stem,
                            'size': f.stat().st_size,
                            'modified': f.stat().st_mtime
                        }
                        for f in quote_path.glob('*.parquet')
                    ]
                
                # List trade data files
                trade_path = catalog_path / 'trade'
                if trade_path.exists():
                    result['trades'] = [
                        {
                            'filename': f.name,
                            'symbol': f.stem,
                            'size': f.stat().st_size,
                            'modified': f.stat().st_mtime
                        }
                        for f in trade_path.glob('*.parquet')
                    ]
            
            # Mock some data for now if empty
            if not any(result.values()):
                result['bars'] = [
                    {'filename': 'NVDA-1-MINUTE.parquet', 'symbol': 'NVDA', 'timeframe': '1-MINUTE', 'size': 12485760, 'modified': 1723325000},
                    {'filename': 'TSLA-1-MINUTE.parquet', 'symbol': 'TSLA', 'timeframe': '1-MINUTE', 'size': 8945632, 'modified': 1723324000},
                    {'filename': 'SPY-1-DAY.parquet', 'symbol': 'SPY', 'timeframe': '1-DAY', 'size': 2457600, 'modified': 1723320000}
                ]
                result['signals'] = [
                    {'filename': 'momentum_signals.parquet', 'size': 125829120, 'modified': 1723325000},
                    {'filename': 'ml_features_v2.parquet', 'size': 398458880, 'modified': 1723324000}
                ]
                result['backtests'] = [
                    {'filename': 'ema_cross_results.parquet', 'size': 5242880, 'modified': 1723320000},
                    {'filename': 'rsi_meanrev_results.parquet', 'size': 4194304, 'modified': 1723310000}
                ]
            
            return result
            
        except Exception as e:
            raise Exception(f"Failed to list catalog data: {str(e)}")
    
    def close(self):
        """Close connections"""
        if hasattr(self.data_manager, 'close'):
            self.data_manager.close()